python/cudf/cudf/core/frame.py

```diff
@@ -1329,7 +1298,40 @@ def _merge(
                 )
         gdf_result._data = to_frame_data
 
-        return self.__class__._from_table(gdf_result)
+        to_return = self.__class__._from_table(gdf_result)
+
+        # If sort=True, Pandas would sort on the key columns in the
+        # same order as given in 'on'. If the indices are used as
+        # keys, the index will be sorted. If one index is specified,
+        # the key column on the other side will be used to sort.
+        # If no index is specified, return a new RangeIndex
+        if sort:
+            to_sort = self.__class__()
+            if left_index and right_index:
+                by = list(to_return._index._data.columns)
+                if left_on and right_on:
+                    by += list(to_return[left_on]._data.columns)
+            elif left_index:
+                by = list(to_return[right_on]._data.columns)
+            elif right_index:
+                by = list(to_return[left_on]._data.columns)
+            else:
+                # left_on == right_on, or different names but same columns
+                # in both cases we can sort by either
+                by = list(to_return[left_on]._data.columns)
+            for i, col in enumerate(by):
+                to_sort[i] = col
+            inds = to_sort.argsort()
+            for col in to_return.columns:
+                to_return._data[col] = to_return._data[col].take(inds)
```

Comment:
Can we do a gather all at once instead of column by column? This is optimized in the C++.

Reply:
Should be updated. The disadvantage here is that we end up materializing and then sorting a RangeIndex, which we then potentially throw away. There might be a way to refactor that out though, I am looking into it.

Reply:
If there's not a way to ignore the index in a gather operation I think that would be an appreciated additional option to avoid materialization.